Add option and logic to disable additional out-of-zone data.

greg-bock opened the pull request:
NSD follows CNAME, DNAME, and possibly other data outside of the queried zone if the target zone is also configured in NSD. This may not always be desirable. This PR adds a configuration option and logic to disable this behavior. The option defaults to the current behavior. 

nsd.conf
```
server:
        ip-address: 127.0.0.1
        database: ""

remote-control:
        control-enable: yes
        control-interface: 127.0.0.1

zone:
        name: foo.test
        zonefile: /etc/nsd/foo.test.zone
zone:
        name: bar.test
        zonefile: /etc/nsd/bar.test.zone
```

foo.test.zone
```
@ IN SOA ns0.foo.test. username.foo.test. 0 0 0 0 0
cname-zone IN CNAME zone
cname-local IN CNAME  local.bar.test.
cname-remote IN CNAME remote.biz.test.
zone A 127.0.0.1
```
bar.test.zone
```
@ IN SOA ns0.bar.test. username.bar.test. 0 0 0 0 0
local A 127.0.0.2
```

Output:
```
[root@nsd-testing nsd]# dig +short @127.0.0.1 cname-zone.foo.test
zone.foo.test.
127.0.0.1
[root@nsd-testing nsd]# dig +short @127.0.0.1 cname-local.foo.test
local.bar.test.
127.0.0.2
[root@nsd-testing nsd]# dig +short @127.0.0.1 cname-remote.foo.test
remote.biz.test.
```

With the PR changes and additional-from-auth set to no:
```
root@nsd-testing nsd]# dig +short @127.0.0.1 cname-zone.foo.test
zone.foo.test.
127.0.0.1
[root@nsd-testing nsd]# dig +short @127.0.0.1 cname-local.foo.test
local.bar.test.
[root@nsd-testing nsd]# dig +short @127.0.0.1 cname-remote.foo.test
remote.biz.test.
```
I'm open to suggestions on the option name (I took this one from bind although it appears to be obsoleted now). I also contemplated allowing zones/patterns to override the global setting but wasn't sure if calling zone_options_find in the query path was a good idea.

Commit:
Add option and logic to not return additional information from disparate
zones.

NSD currently returns additional information like cname resolution from
zones other than the query apex zone if the target zone is local. This
is not always desirable.

Commit:
Fix typo.

Commit:
Fix formatting.

k0ekk0ek commented:
Hi @greg-bock! Thanks for the PR. Most of the code looks good, just two minor remarks.

1. I'd prefer you'd use `dname_compare()` instead of testing for pointer equality. While that'll work just fine for now, in the future there could be modifications that would invalidate the check.
2. The name `additional-from-auth` doesn't hold much value in NSD as it's an authoritative-only nameserver (bind had `additional-from-cache` too). Apart from that, as you stated, the option has been removed from bind in 9.12, which may cause some confusion. How about `confine-to-zone: <yes|no>`? That way the option can also just be initialized to `0`.

I'll happily accept the changes once the issues have been resolved.

greg-bock commented:
I'll make the changes shortly. Thanks for the feedback!

Commit:
Rename additional-from-auth to confine-to-zone.

Rename option.
Fix logic and man page as the boolean meaning has flipped.
Use dname_compare instead of direct comparison.

greg-bock commented:
@k0ekk0ek let me know if you want me to squash this into a single commit and force push the branch to clean up history.

k0ekk0ek commented:
@greg-bock, sorry for the delay. I wanted to give the other committers a chance to react to the proposed changes. I'm assuming they're fine with the changes, so I'll go ahead and merge this. Squashing won't be necessary, GitHub allows me to squash the changes and combine them in a single commit. Thanks again!

greg-bock commented:
Thanks!
